app/src/components/GameRoutes.tsx

```diff
@@ -15,23 +27,82 @@ function CurrentPlayerProvider(props: {
   joinCode: string
   children: React.ReactNode
 }) {
-  const { data, loading } = useCurrentPlayerQuery({
+  const currentAuth = React.useContext(CurrentAuthContext)
+
+  const [currentPlayer, setCurrentPlayer] = React.useState<
+    CurrentPlayerQuery["players"][0] | null
+  >(null)
+  const [redirectRoute, setRedirectRoute] = React.useState("")
+
+  // https://github.com/apollographql/react-apollo/issues/3505#issuecomment-568112061
+  const [skipCallback, setSkipCallback] = React.useState(false)
+
+  const [joinGame] = useJoinGameMutation()
+  const [signUp] = useLazyQuery(GameByJoinCodeDocument, {
+    variables: { joinCode: props.joinCode.toLocaleUpperCase() },
+    onCompleted: async data => {
+      if (skipCallback) {
+        return
+      }
+      if (data && data.games[0]) {
+        const registration = await joinGame({
+          variables: {
+            gameId: data.games[0].id,
+            clientUuid: clientUuid()
+          }
+        })
+        if (registration.data?.joinGame) {
+          await currentAuth.setJwtToken(registration.data.joinGame.jwt_token)
+        }
+        setSkipCallback(true)
+      } else {
+        setRedirectRoute(routes.game.root)
+      }
+    },
+    onError: _ => {
+      setRedirectRoute(routes.game.root)
+    }
+  })
+
+  const [
+    signIn,
```

Comment:
so "sign in" essentially means "look for a jwt in local storage and see if there is a current player associated with it"?

what stops us from pre-maturely redirecting when `loadingSignIn` is true?

Reply:
tho only place where redirectRoute is set is in an onCompletion or onError branch - so you know loadingSignIn will be false I think

Does that make sense? It could be better protected for more defenesive programming

Reply:
e.g.
```
  if (!loadingSignIn && redirectRoute) {
    return <Redirect to={routes.root}></Redirect>
  }
```

Reply:
@wtachau this is a good line of thought coming back to it... the issue is a redirect of the page, and this is the only thing that redirects haha... will merge and dig into it later! thanks for flagging.